### PR TITLE
WIP: adding possibility to have 'online-cmn' on input of 'nnet3' models

### PR DESCRIPTION
**Pull request by vesis84**
Links: This code is related to issue #3461 (specification of code changes), and PR #3439 that will add the support of online-cmvn frontend in the online2 C++ codebase.

Characteristics of this PR:
- online-cmvn can be used with or without ivector features,
- it is demonstrated for the 'rm' databse,
- the online-cmn is supported for both 'chain' or 'cross-entropy' nnet3 models,
- the WER's for chain RM models were identical, see below.

- without cmn,
%WER 1.60 [ 201 / 12533, 17 ins, 34 del, 150 sub ] exp/chain/tdnn_5o/decode/wer_3_0.0
%WER 8.11 [ 1016 / 12533, 76 ins, 140 del, 800 sub ] exp/chain/tdnn_5o/decode_ug/wer_7_0.0
- with online-cmn,
%WER 1.57 [ 197 / 12533, 21 ins, 29 del, 147 sub ] exp/chain/tdnn_5o_online-cmn/decode/wer_2_0.5
%WER 8.11 [ 1017 / 12533, 59 ins, 160 del, 798 sub ] exp/chain/tdnn_5o_online-cmn/decode_ug/wer_8_0.0

Dan, any comments, suggestions? 
What should be the next 'egs' I should demonstrate this on? 'wsj' or something else?
Are you happy with the way how the 'online-cmvn' support is added? 

Not expecting this PR to merge yet, first we should add a demo for another recipe 
and if someone could help with testing, that would be fantastic...

Thanks,
K.

### CODE REVIEW

File: egs/wsj/s5/steps/libs/nnet3/train/common.py [hunk 1]
```diff
@@ -320,7 +320,7 @@ def halve_minibatch_size_str(minibatch_size_str):
 
 def copy_egs_properties_to_exp_dir(egs_dir, dir):
     try:
-        for file in ['cmvn_opts', 'splice_opts', 'info/final.ie.id', 'final.mat']:
+        for file in ['cmvn_opts', 'splice_opts', 'info/final.ie.id', 'final.mat', 'global_cmvn.stats', 'online_cmvn_frontend']:
```
Comment (danpovey) [inline egs/wsj/s5/steps/libs/nnet3/train/common.py]: please split line

File: egs/wsj/s5/steps/nnet3/chain/get_egs.sh [hunk 2]
```diff
@@ -65,6 +65,7 @@ online_ivector_dir=  # can be used if we are including speaker information as iV
 cmvn_opts=  # can be used for specifying CMVN options, if feature type is not lda (if lda,
             # it doesn't make sense to use different options than were used as input to the
             # LDA transform).  This is used to turn off CMVN in the online-nnet experiments.
+online_cmvn_frontend=false # set to 'true' to activate online cmvn,
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/chain/get_egs.sh]: clarify: "(for the regular features; separate from the online CMVN used for the ivectors)"
and please remove the _frontend; I think online_cmvn is sufficient.

File: egs/wsj/s5/steps/online/nnet2/train_ivector_extractor.sh [hunk 3]
```diff
@@ -41,6 +41,7 @@ num_processes=4 # each job runs this many processes, each with --num-threads thr
 cmd="run.pl"
 stage=-4
 ivector_dim=100 # dimension of the extracted i-vector
+online_cmvn=false # apply online-cmvn on i-vector input features, uses the configuration from UBM,
```
Comment (danpovey) [inline egs/wsj/s5/steps/online/nnet2/train_ivector_extractor.sh]: .. I think here, though, calling the variable online_cmvn_iextractor might be clearer

File: egs/wsj/s5/steps/online/nnet2/train_ivector_extractor.sh [hunk 4]
```diff
@@ -115,6 +116,10 @@ splice_opts=$(cat $srcdir/splice_opts)
 gmm_feats="ark,s,cs:apply-cmvn-online --config=$dir/online_cmvn.conf $dir/global_cmvn.stats scp:$sdata/JOB/feats.scp ark:- | splice-feats $splice_opts ark:- ark:- | transform-feats $dir/final.mat ark:- ark:- | subsample-feats --n=$subsample ark:- ark:- |"
 feats="ark,s,cs:splice-feats $splice_opts scp:$sdata/JOB/feats.scp ark:- | transform-feats $dir/final.mat ark:- ark:- | subsample-feats --n=$subsample ark:- ark:- |"
 
+## This adds online-cmvn in $feats, upon request (configuration taken from UBM),
+## ('online_cmvn_iextractor' marks that we added online_cmvn)
+[ $online_cmvn ] && { feats="$gmm_feats"; touch $dir/online_cmvn_iextractor; }
```
Comment (danpovey) [inline egs/wsj/s5/steps/online/nnet2/train_ivector_extractor.sh]: if  the variable is false, I think you should remove it just in case the dir was used before. e.g. `rm foo 2>/dev/null || true`.  Also please don't use brackets around $online_cmvn; it's true or false so it doesn't need them, and they are confusing.

File: src/online2/online-ivector-feature.cc [hunk 5]
```diff
@@ -225,7 +226,14 @@ void OnlineIvectorFeature::UpdateStatsForFrames(
         posterior[j].second *= info_.posterior_scale * weight;
     }
   }
-  lda_->GetFrames(frames, &feats);  // get features without CMN.
+
+  // Do we select features with online cmvn?
```
Comment (danpovey) [inline src/online2/online-ivector-feature.cc]: this comment is confusing to me, it sounds like the person coding it was not sure.  better to remove.

**Comment by vesis84** [comment]
Okay, all remarks resolved. 

The next step is to test online-cmvn on : mini_librispeech next, then maybe WSJ or tedlium.
This will be a part of this PR.

**Comment by vesis84** [comment]
The code for: mini_librispeech, wsj, tedlium is in place. waiting for the results now...

**Comment by vesis84** [comment]
I have the mini_librispeech results:
```
tglarge tgsmall
8.13  11.58  tdnn1i_sp, no cmvn,
7.61  11.15  tdnn1j_sp, online-cmn
```
The online-cmn helps quite a bit... Wsj and tedlium is still running...

**Comment by danpovey** [comment]
Great news!
Would you mind re-doing these experiments on top of the PR that I merged
just this minute, from Desh, which was about replacing the LDA with delta +
delta-delta?  It's just a simplification.  But I'd like to make the
online-cmn the next default script, so it should be combined with the
existing changes.

Dan

**Comment by danpovey** [comment]
@HuangZiliAndy  do you have time to test this out?  I know you don't normally use this type of setup, but it wouldn't hurt you to try.  You could ask help from some of the others.
Outstanding tasks:
Fix conflicts.  You'd probably want to apply the same changes that Karel made in (say) i->j, and apply them on top of the current 'j' script in the repo, so it wouldn't just be a case of renaming j to k.  And then, of course, testing on RM and WSJ, and hopefully also in another setup, like tedlium s5_r2 or AMI.
Obviously this will have to wait till the cluster is back up.  You might want to ask help from those more familiar with these setups.

@vimalmanohar, can you please comment on how this differs from whatever you were testing a couple months ago?

**Comment by vesis84** [comment]
I am still building the systems. Our cluster is not so reliable these days with so many GPU jobs.

Which of the recipes should be run with delta-delta alignments? 
Only the mini_librispeech or also the other recipes?

K.

**Comment by HuangZiliAndy** [comment]
Hi Dan,

    I am glad to take a look. Thanks

Best
Zili

获取 Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by vesis84** [comment]
And here are the results without the 'Desh' update:

mini_librispeech (5.3h)
```
tglarge tgsmall
8.13    11.58  tdnn1i_sp, no cmvn,
7.61    11.15  tdnn1j_sp, online-cmn
=> clearly better
```

wsj (81.3h)
```
# Chain results (tdnn-f, ivectors)
# no-cmvn (local/chain/run_tdnn.sh)
%WER 4.96 [ 408 / 8234, 47 ins, 59 del, 302 sub ] exp/chain/tdnn1g_sp/decode_bd_tgpr_dev93/wer_11_0.0
%WER 2.80 [ 158 / 5643, 16 ins, 7 del, 135 sub ] exp/chain/tdnn1g_sp/decode_bd_tgpr_eval92/wer_10_1.0
%WER 6.89 [ 567 / 8234, 77 ins, 111 del, 379 sub ] exp/chain/tdnn1g_sp/decode_tgpr_dev93/wer_12_0.0
%WER 4.75 [ 268 / 5643, 36 ins, 55 del, 177 sub ] exp/chain/tdnn1g_sp/decode_tgpr_eval92/wer_14_1.0
# online-cmvn (local/chain/run_tdnn_online_cmn.sh)
%WER 4.93 [ 406 / 8234, 36 ins, 68 del, 302 sub ] exp/chain_online_cmn/tdnn1h_sp/decode_bd_tgpr_dev93/wer_12_0.5
%WER 2.87 [ 162 / 5643, 21 ins, 4 del, 137 sub ] exp/chain_online_cmn/tdnn1h_sp/decode_bd_tgpr_eval92/wer_9_0.5
%WER 6.97 [ 574 / 8234, 77 ins, 106 del, 391 sub ] exp/chain_online_cmn/tdnn1h_sp/decode_tgpr_dev93/wer_11_0.5
%WER 4.78 [ 270 / 5643, 35 ins, 51 del, 184 sub ] exp/chain_online_cmn/tdnn1h_sp/decode_tgpr_eval92/wer_13_1.0
=> no change or little worse
```

tedlium (431h)
```
# no cmvn (only i-vector UBM with online cmn),
%WER 8.32 [ 1480 / 17783, 253 ins, 292 del, 935 sub ] exp/chain_cleaned/tdnn1c_sp/decode_dev/wer_9
%WER 7.63 [ 1357 / 17783, 240 ins, 303 del, 814 sub ] exp/chain_cleaned/tdnn1c_sp/decode_dev_rescore/wer_10
%WER 8.44 [ 2321 / 27500, 351 ins, 553 del, 1417 sub ] exp/chain_cleaned/tdnn1c_sp/decode_test/wer_9
%WER 7.84 [ 2155 / 27500, 311 ins, 577 del, 1267 sub ] exp/chain_cleaned/tdnn1c_sp/decode_test_rescore/wer_10
# online-cmn,
%WER 8.22 [ 1461 / 17783, 236 ins, 296 del, 929 sub ] exp/chain_cleaned_1d/tdnn1d_sp/decode_dev/wer_10
%WER 7.60 [ 1352 / 17783, 257 ins, 252 del, 843 sub ] exp/chain_cleaned_1d/tdnn1d_sp/decode_dev_rescore/wer_9
%WER 8.45 [ 2324 / 27500, 347 ins, 549 del, 1428 sub ] exp/chain_cleaned_1d/tdnn1d_sp/decode_test/wer_9
%WER 7.73 [ 2127 / 27500, 285 ins, 623 del, 1219 sub ] exp/chain_cleaned_1d/tdnn1d_sp/decode_test_rescore/wer_11
=> no change or little better
```

All the systems have speed perturbation and volume perturbation.

The way the script work is: 'local/chain/run_tdnn_online_cmn.sh' assumes that the 'local/chain/run_tdnn.sh' was run already... (features and GMM alignments are already prepared). If not, it is still possible to run 'local/chain/run_tdnn_online_cmn.sh --stage 0 --skip-alignment false' (data preparation fails if into same directories for the 2nd time). Is it okay like that? (I can add a comment if necessary).

Dan, what should be the next step to accept these build-scripts?
(the porting to online2 will be resolved in another PR)
K.

File: egs/mini_librispeech/s5/local/chain/tuning/run_tdnn_1j.sh [hunk 6]
```diff
@@ -0,0 +1,300 @@
+#!/bin/bash
+
```
Comment (danpovey) [inline egs/mini_librispeech/s5/local/chain/tuning/run_tdnn_1j.sh]: Please don't have a separate link to this (RE `online_cmvn`.)-- I'd like to just advance the regular link. The idea is that it just becomes the new TDNN recipe.  (If we add a new suffix each time we change something it quickly becomes unmanageable).
The other task is to resolve conflicts.  Essentially this means moving forward j -> k and applying whatever changes I had in i->j.

File: egs/wsj/s5/steps/nnet3/chain/get_egs.sh [hunk 7]
```diff
@@ -65,6 +65,9 @@ online_ivector_dir=  # can be used if we are including speaker information as iV
 cmvn_opts=  # can be used for specifying CMVN options, if feature type is not lda (if lda,
             # it doesn't make sense to use different options than were used as input to the
             # LDA transform).  This is used to turn off CMVN in the online-nnet experiments.
+online_cmvn=false # set to 'true' to activate online_cmvn on the nnet3 input,
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/chain/get_egs.sh]: Karel,  can you clarify in this comment that this uses (and I assume requires?) the config in conf/online_cmn.conf?

Comment (danpovey) [reply egs/wsj/s5/steps/nnet3/chain/get_egs.sh]: and if it uses that config, I think the normal practice would be to make that configurable?  So we dont have difficulty to change those configuration values when running multiple experiments in the same dir?

File: egs/wsj/s5/steps/nnet3/decode.sh [hunk 8]
```diff
@@ -96,11 +96,14 @@ mkdir -p $dir/log
 [[ -d $sdata && $data/feats.scp -ot $sdata ]] || split_data.sh $data $nj || exit 1;
 echo $nj > $dir/num_jobs
 
-
 ## Set up features.
+[ -f $srcdir/online_cmvn ] && online_cmvn=true || online_cmvn=false
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/decode.sh]: this is a little hard to parse.  Would you mind just making it a regular if-statement?  you can fit it on 2 lines.

File: egs/mini_librispeech/s5/cmd.sh [hunk 9]
```diff
@@ -13,3 +13,10 @@
 export train_cmd="queue.pl --mem 2G"
 export decode_cmd="queue.pl --mem 4G"
 export mkgraph_cmd="queue.pl --mem 8G"
+
+if [ "$(hostname -d)" == "fit.vutbr.cz" ]; then
```
Comment (danpovey) [inline egs/mini_librispeech/s5/cmd.sh]: I'd prefer if you don't check this part in, and just keep it local.

File: egs/wsj/s5/steps/nnet3/chain/get_egs.sh (hunk 7 above)
Comment (vesis84) [reply egs/wsj/s5/steps/nnet3/chain/get_egs.sh]: Hi, I don't refer the 'conf/online_cmn.conf' directly in this script. Instead, I am passing it in $cmvn_opts through train.py as --cmvn-opts="--config=conf/online_cmn.conf". So it is configurable from outside and it re-uses former mechanism for configuring the cmvn.

**Comment by vesis84** [comment]
Okay, i have updated the PR so it is on top of HEAD and also the online-cmn recipes contain the delta-delta front-end. All the remarks are incorporated.

@danpovey please check the code again, meanwhile I'll run the experiments and update the results 
(I'll put them input the local/chain/tuning/run_tdnn*.sh with the comparison to 'previous-letter' recipe)

K.

**Comment by danpovey** [comment]
Thanks!
Perhaps you can suggest a specific setup that you might like @HuangZiliAndy to run?

**Comment by danpovey** [comment]
Actually, @HuangZiliAndy, I think the AMI setup with distant microphones (SDM) would be a good one.  When the grid is up properly :-(

I am not going to get too personally involved in fixing it; I want to let the CLSP people who agitated to get me fired feel the consequences of that.

**Comment by vesis84** [comment]
Okay, i finished the changes and updated the results. 
Perhaps, I should not sqash the commits as @HuangZiliAndy might be using my branch...

**Comment by vesis84** [comment]
Should I also update the symlinks? local/chain/train_tdnn.sh to point to my scripts?

**Comment by vesis84** [comment]
Okay, I'll be on vacation whole next week. Either it will wait a bit or you can finalize this... K.

**Comment by danpovey** [comment]
@vesis84 if your new scripts are as good, or almost as good, as the baseline I'd say yes, update the symlinks.
(The reason I say even if "almost as good", is that CMN'ed systems are expected to be more robust.)

**Comment by vesis84** [comment]
Hi, coming back after the vacation. 

Yes, I would say it qualifies for being "almost as good", the updated results are here:

mini_librispeech
```
# local/chain/compare_wer.sh --online exp/chain/tdnn1j_sp exp/chain_online_cmn/tdnn1k_sp
# System                tdnn1j_sp tdnn1k_sp
#WER dev_clean_2 (tgsmall)      11.25     10.99
#WER dev_clean_2 (tglarge)       7.72      7.54
# Final train prob        -0.0632   -0.0623
# Final valid prob        -0.0792   -0.0803
# Final train prob (xent)   -1.4499   -1.4396
# Final valid prob (xent)   -1.5643   -1.5628
# Num-params                 5210944   5210944
```

tedlium
```
# local/chain/compare_wer_general.sh exp/chain_cleaned/tdnn1c_sp exp/chain_cleaned_1d/tdnn1d_sp
# System                tdnn1c_sp tdnn1d_sp
# WER on dev(orig)           8.32      8.22
# WER on dev(rescored)       7.63      7.60
# WER on test(orig)          8.44      8.45
# WER on test(rescored)      7.84      7.73
# Final train prob        -0.0688   -0.0677
# Final valid prob        -0.0826   -0.0847
# Final train prob (xent)   -0.9842   -0.9814
# Final valid prob (xent)   -1.0976   -1.0930
# Num-params                 9476304   9476304
```

wsj
```
# local/chain/compare_wer.sh exp/chain/tdnn1h_sp exp/chain_online_cmn/tdnn1i_sp
# System                tdnn1h_sp tdnn1i_sp
#WER dev93 (tgpr)                6.89      6.66
#WER dev93 (tg)                  6.63      6.56
#WER dev93 (big-dict,tgpr)       4.96      4.74
#WER dev93 (big-dict,fg)         4.53      4.44
#WER eval92 (tgpr)               4.68      4.89
#WER eval92 (tg)                 4.32      4.43
#WER eval92 (big-dict,tgpr)      2.69      2.91
#WER eval92 (big-dict,fg)        2.34      2.37
# Final train prob        -0.0442   -0.0439
# Final valid prob        -0.0537   -0.0529
# Final train prob (xent)   -0.6548   -0.6581
# Final valid prob (xent)   -0.7324   -0.7300
# Num-params                 8349232   8349232
```

So, I will update the symlinks now...

**Comment by vesis84** [comment]
@HuangZiliAndy: Do you have any news with the AMI results?

**Comment by vesis84** [comment]
Yes, the change is done and is ready to merge. I AM SPARTACUS !!!

**Comment by HuangZiliAndy** [comment]
Hi, I experimented on AMI and SWBD. It seems that the performance declined. Thanks!
[https://docs.google.com/spreadsheets/d/1RmYSa77IoIdaFOg1FJL3EUBDqBQ1B2nImITvR3ssBw8/edit?usp=sharing](url)

**Comment by HuangZiliAndy** [comment]
Okay, I will run this on AMI. Thanks!

Karel Vesely <notifications@github.com> 于2019年9月23日周一 上午10:46写道：

**Comment by vesis84** [comment]
And here are my results, the change in branch `https://github.com/vesis84/kaldi/tree/online_cmvn_consistency` 
seems to fix the problem...

But in the same time, it leads to slightly different behavior when training without online-cmvn.
The transfer of cmvn-stats for 'ubm' input features was not there before during training, but the transfer was done in the online2 code...

@danpovey Would that be okay to add the online-cmvn-stats transfer into 'master'?

# wsj

## no transfer
```
# local/chain/compare_wer.sh exp/chain/tdnn1h_sp exp/chain_online_cmn/tdnn1i_sp
# System                tdnn1h_sp tdnn1i_sp
#WER dev93 (tgpr)                6.89      6.90
#WER dev93 (tg)                  6.63      6.73
#WER dev93 (big-dict,tgpr)       4.96      4.91
#WER dev93 (big-dict,fg)         4.53      4.44
#WER eval92 (tgpr)               4.68      4.77
#WER eval92 (tg)                 4.32      4.36
#WER eval92 (big-dict,tgpr)      2.69      2.85
#WER eval92 (big-dict,fg)        2.34      2.36
```

## cmvn-stats transfer in ubm,iextractor
```
# local/chain/compare_wer.sh exp/chain/tdnn1h_sp exp/chain_online_cmn/tdnn1i_sp
# System                tdnn1h_sp tdnn1i_sp
#WER dev93 (tgpr)                6.89      6.64
#WER dev93 (tg)                  6.63      6.51
#WER dev93 (big-dict,tgpr)       4.96      4.63
#WER dev93 (big-dict,fg)         4.53      4.44
#WER eval92 (tgpr)               4.68      4.75
#WER eval92 (tg)                 4.32      4.29
#WER eval92 (big-dict,tgpr)      2.69      2.69
#WER eval92 (big-dict,fg)        2.34      2.27
```
# tedlium/s5_r3

## no transfer
```
# local/chain/compare_wer_general.sh exp/chain_cleaned/tdnn1c_sp exp/chain_cleaned_1d/tdnn1d_sp
# System                tdnn1c_sp tdnn1d_sp
# WER on dev(orig)           8.32      8.50
# WER on dev(rescored)       7.63      7.91
# WER on test(orig)          8.44      8.39
# WER on test(rescored)      7.84      7.88
```

## cmvn-stats transfer in ubm,iextractor
```
# System                tdnn1c_sp tdnn1d_sp
# WER on dev(orig)           8.32      8.37
# WER on dev(rescored)       7.63      7.61
# WER on test(orig)          8.44      8.36
# WER on test(rescored)      7.84      7.89
```

**Comment by danpovey** [comment]
@vesis84  Yes it would be great to add that.. please make a PR!

**Comment by HuangZiliAndy** [comment]
Hi Karel,

    Seems the performance is almost the same with and w/o the change on
AMI.


https://docs.google.com/spreadsheets/d/1RmYSa77IoIdaFOg1FJL3EUBDqBQ1B2nImITvR3ssBw8/edit?usp=sharing

Thanks
Zili

Daniel Povey <notifications@github.com> 于2019年9月24日周二 下午12:26写道：

**Comment by danpovey** [comment]
Thanks!  We could perhaps leave this up just as a reminder/note that online
CMVN doesn't seem to work on AMI.
Which is not surprising given that it could have background speech.

But Zili, can you try to somehow insert your table into the conversation
directly rather than relying on the link?  (I'm not sure whether copy/paste
will do anything useful.. anyway it doesn't have to be pretty.) . The link
will die one day.

**Comment by HuangZiliAndy** [comment]
The performance on AMI is not satisfying. The WER on AMI test set increased from 35.4% to 36.2%.

| system   | 1j   | 1k   | 1k(consistency) |
|----------|------|------|-----------------|
| AMI dev  | 32.0  | 32.9 | 32.9            |
| AMI test | 35.4 | 36.2 | 36.4            |
